core/src/main/scalajs/sttp/client3/AbstractFetchBackend.scala

```diff
@@ -240,63 +249,104 @@ abstract class AbstractFetchBackend[F[_], S <: Streams[S], P](
       .order(original.order)
   }
 
-  protected def handleStreamBody(s: streams.BinaryStream): F[js.UndefOr[BodyInit]]
+  private def sendWebSocket[T, R >: PE](request: Request[T, R]): F[Response[T]] = {
+    val queue = new JSSimpleQueue[F, WebSocketEvent](webSocketTimeout)
+    val ws = new JSWebSocket(request.uri.toString)
+    ws.binaryType = BinaryType
 
-  private lazy val bodyFromResponseAs =
-    new BodyFromResponseAs[F, FetchResponse, Nothing, streams.BinaryStream] {
-      override protected def withReplayableBody(
-          response: FetchResponse,
-          replayableBody: Either[Array[Byte], SttpFile]
-      ): F[FetchResponse] = {
-        val bytes = replayableBody match {
-          case Left(byteArray) => byteArray
-          case Right(file)     => throw new IllegalArgumentException("Replayable file bodies are not supported")
-        }
-        new FetchResponse(bytes.toTypedArray.asInstanceOf[BodyInit], response.asInstanceOf[ResponseInit]).unit
-      }
+    ws.onopen = (_: Event) => queue.offer(WebSocketEvent.Open())
+    ws.onmessage = (event: MessageEvent) => queue.offer(toWebSocketEvent(event))
+    ws.onerror = (_: Event) => queue.offer(WebSocketEvent.Error(new WebSocketErrorException))
+    ws.onclose = (event: CloseEvent) => queue.offer(toWebSocketEvent(event))
 
-      override protected def regularIgnore(response: FetchResponse): F[Unit] =
-        transformPromise(response.arrayBuffer()).map(_ => ())
+    val webSocket = WebSocketImpl.newJSCoupledWebSocket(ws, queue)
 
-      override protected def regularAsByteArray(response: FetchResponse): F[Array[Byte]] = responseToByteArray(response)
+    val isOpen = JSTimeout[Unit] {
+      Either.cond(ws.readyState == OpenState, (), new ReadException(request, new WebSocketTimeoutException))
+    }(webSocketTimeout)
 
-      override protected def regularAsFile(response: FetchResponse, file: SttpFile): F[SttpFile] = {
-        transformPromise(response.arrayBuffer())
-          .map { ab =>
-            SttpFile.fromDomFile(
-              new DomFile(
-                Array(ab.asInstanceOf[js.Any]).toJSArray,
-                file.name,
-                FilePropertyBag(`type` = file.toDomFile.`type`)
-              )
-            )
-          }
-      }
+    fromFuture(isOpen).flatMap { _ =>
+      bodyFromResponseAs
+        .apply(request.response, ResponseMetadata(StatusCode.Ok, "", request.headers), Right(webSocket))
+        .map(Response.ok)
+    }
+  }
 
-      override protected def regularAsStream(response: FetchResponse): F[(streams.BinaryStream, () => F[Unit])] =
-        handleResponseAsStream(response)
+  def webSocketTimeout: FiniteDuration = 1.second
```

Comment:
does this mean that if a message isn't received every second you'll get an error?

Reply:
No it defines timeout for waiting for web socket to open, and timeout for each `ws.receive()` call